snapcraft/internal/lifecycle.py

```diff
@@ -442,32 +438,81 @@ def _cleanup_common_directories(config, project_options):
-    if (max_index < common.COMMAND_ORDER.index('pull') and
-            os.path.exists(project_options.parts_dir) and not
-            os.path.exists(project_options.local_plugins_dir)):
-        logger.info('Cleaning up parts directory')
-        shutil.rmtree(project_options.parts_dir)
+    with contextlib.suppress(IndexError):
+        _cleanup_common_directories_for_step(
+            common.COMMAND_ORDER[max_index+1], project_options)
+
 
-    # If no parts have been staged, remove staging area.
-    should_remove_stagedir = max_index < common.COMMAND_ORDER.index('stage')
-    if should_remove_stagedir and os.path.exists(project_options.stage_dir):
-        logger.info('Cleaning up staging area')
-        shutil.rmtree(project_options.stage_dir)
+def _cleanup_common_directories_for_step(step, project_options, parts=None):
+    if not parts:
+        parts = []
 
-    # If no parts have been primed, remove snapping area.
-    should_remove_snapdir = max_index < common.COMMAND_ORDER.index('prime')
-    if should_remove_snapdir and os.path.exists(project_options.snap_dir):
-        logger.info('Cleaning up snapping area')
-        shutil.rmtree(project_options.snap_dir)
+    index = common.COMMAND_ORDER.index(step)
+
+    if index <= common.COMMAND_ORDER.index('prime'):
+        # Remove the priming area.
+        _cleanup_common(
+            project_options.snap_dir, 'prime', 'Cleaning up priming area',
+            parts)
+
+    if index <= common.COMMAND_ORDER.index('stage'):
+        # Remove the staging area.
+        _cleanup_common(
+            project_options.stage_dir, 'stage', 'Cleaning up staging area',
+            parts)
+
+    if index <= common.COMMAND_ORDER.index('pull'):
+        # Remove the parts directory (but leave local plugins alone).
+        _cleanup_parts_dir(
+            project_options.parts_dir, project_options.local_plugins_dir,
+            parts)
+
+    _remove_directory_if_empty(project_options.snap_dir)
+    _remove_directory_if_empty(project_options.stage_dir)
+    _remove_directory_if_empty(project_options.parts_dir)
+
+
+def _cleanup_common(directory, step, message, parts):
+    if os.path.isdir(directory):
+        logger.info(message)
+        shutil.rmtree(directory)
+    for part in parts:
+        part.mark_cleaned(step)
+
+
+def _cleanup_parts_dir(parts_dir, local_plugins_dir, parts):
+    if os.path.exists(parts_dir):
+        logger.info('Cleaning up parts directory')
+        for subdirectory in os.listdir(parts_dir):
+            path = os.path.join(parts_dir, subdirectory)
+            if path != local_plugins_dir:
+                try:
+                    shutil.rmtree(path)
+                except NotADirectoryError:
+                    os.remove(path)
+    for part in parts:
+        part.mark_cleaned('build')
+        part.mark_cleaned('pull')
 
 
 def clean(project_options, parts, step=None):
+    if not step:
+        step = 'pull'
+
+    if not parts and step == 'pull':
```

Comment:
it feels like what you want here is
`if not parts and not step` and keep `step` as None? If we ever add a new pre-pull step this will be a point of failure, I get that there may be others.